Index: dlldrvobj.h
===================================================================
--- dlldrvobj.h	(revision 3328)
+++ dlldrvobj.h	(revision 3343)
@@ -31,7 +31,7 @@
 
 #define DLL_MAIN_VERSION					1
 #define DLL_MINOR_VERSION					0
-#define DLL_SUB_VERSION				        6	
+#define DLL_SUB_VERSION				        7	
 
 // This is the vendor string - Change to your own value
 #define CANAL_DLL_VENDOR "edevices, Lithuania, http://www.8devices.com"
@@ -40,6 +40,7 @@
 // Max number of open connections
 //#define CANAL_USB2CAN_DRIVER_MAX_OPEN	256
 #define CANAL_USB2CAN_DRIVER_MAX_OPEN	1
+//WARNING! CANAL_USB2CAN_DRIVER_MAX_OPEN will be a problem with using old version with smaller size and a new version with bigger size on the same pc at the same time
 
 /////////////////////////////////////////////////////////////////////////////
 // CDllDrvObj
@@ -92,7 +93,10 @@
 		This is the array with driver objects (max 256 objects
 	*/
 	CUsb2canObj *m_drvObjArray[ CANAL_USB2CAN_DRIVER_MAX_OPEN ];
+	long * lpvMem;      // pointer to shared memory
+	HANDLE hMapObject;  // handle to file mapping
 
+
 	/// Mutex for open/close
 #ifdef WIN32	
 	HANDLE m_objMutex;
