Index: usb2candrv.cpp
===================================================================
--- usb2candrv.cpp	(revision 3328)
+++ usb2candrv.cpp	(revision 3340)
@@ -47,9 +47,15 @@
 	switch( ul_reason_for_call ) {
 
 		case DLL_PROCESS_ATTACH:
-			hThisInstDll = hInstDll;
-			theApp = new CDllDrvObj();
-			theApp->InitInstance();
+			try{
+				hThisInstDll = hInstDll;
+				theApp = new CDllDrvObj();
+				theApp->InitInstance();
+			}
+			catch(...)
+			{
+				return FALSE;//memory exception creating handles
+			}
 			break;
 
 		case DLL_THREAD_ATTACH:
@@ -90,18 +96,18 @@
 	{
 
 			if ( ( h = theApp->addDriverObject(pdrvObj)) <= 0 )
-				{
+			{
 				 h = 0;
                  delete  pdrvObj; 
-				}
+			}
 			else
+			{
+				if ( pdrvObj->open( pDevice, flags ) == FALSE )
 				{
-                  if ( pdrvObj->open( pDevice, flags ) == FALSE )
-					  {
-					   theApp->removeDriverObject( h );
-					   h = 0;
-					  }
-				}		
+				   theApp->removeDriverObject( h );
+				   h = 0;
+				}
+			}		
 	}
 
 	return h;
@@ -371,16 +377,31 @@
 //#define CANAL_MINOR_VERSION					0
 //#define CANAL_SUB_VERSION	
 
-	strcpy( r_str, pdrvObj->getVendorString());
-    strcat( r_str, ";");
+	strcpy_s( r_str, 256, pdrvObj->getVendorString());
+    strcat_s( r_str, 256, ";");
     sprintf_s(tmp_str,sizeof(tmp_str),"%d.%d.%d",CANAL_MAIN_VERSION,CANAL_MINOR_VERSION,CANAL_SUB_VERSION);
-	strcat( r_str,tmp_str );
-	strcat( r_str, ";");
+	strcat_s( r_str, 256, tmp_str );
+	strcat_s( r_str, 256, ";");
     sprintf_s(tmp_str,sizeof(tmp_str),"%d.%d.%d",DLL_MAIN_VERSION,DLL_MINOR_VERSION,DLL_SUB_VERSION);
-	strcat( r_str,tmp_str );
-	strcat( r_str, ";");
-	strcat( r_str, "8devices.com\0");
+	strcat_s( r_str, 256, tmp_str );
+	strcat_s( r_str, 256, ";");
+	strcat_s( r_str, 256, "8devices.com\0");
 
     return (r_str);
 }
 
+
+///////////////////////////////////////////////////////////////////////////////
+// CanalGetReceiveHandle
+//
+
+#ifdef WIN32
+extern "C" HANDLE WINAPI EXPORT CanalGetReceiveHandle( long handle )
+#else
+extern "C" HANDLE CanalGetReceiveHandle( long handle )
+#endif
+{
+	CUsb2canObj *pdrvObj =  theApp->getDriverObject( handle );
+	if ( NULL == pdrvObj ) return 0;
+	return pdrvObj->getReceiveHandle();
+}
