Index: usb2canobj.h
===================================================================
--- usb2canobj.h	(revision 3328)
+++ usb2canobj.h	(revision 3329)
@@ -379,6 +379,13 @@
 	int dataAvailable( void );
 
 	/*!
+		Handle for receive event to know when to call readMsg
+		@return Handle
+	*/
+	HANDLE getReceiveHandle( void ) { return m_receiveDataEvent; }
+
+
+	/*!
 		Get device status
 		@param pCanalStatus Pointer to CANAL status structure
 		@return True on success.
